uji.py

```diff
@@ -1054,6 +1065,152 @@ def upload(self):
+        in_section = False
+        for line in prev_lines[self.cursor_offset + 1:]:
+            if in_section:
+                if '```' in line:
+                    in_section = False
+                else:
+                    # remove existing output
+                    continue
+            if section_index < 0:
+                # find the existing ``` markers
+                if '```' in line:
+                    in_section = True
+                    section_index = len(self.lines) + 1
+                # new section, enforce our markers, right below cursor_offset
+                if self.is_checkbox(line):
+                    section_index = self._insert(self.cursor_offset + 1, '```\n')
+                    self._insert(section_index, '```\n')
+            # rest is going to be forwarded as it was
+            self.lines.append(line)
+
+        # return the index where to store the output of the command
+        return section_index
+
+    def _sanitize(self, line):
+        return line.replace('\t', '    ')
+
+    def execute_command(self):
+        line = self.lines[self.cursor_offset]
+        if not self.is_checkbox(line) or "⚙" not in line:
+            return
+
+        # list of regex to deduce the type of command to be run
+        command_re = {
+                'attach': r'^(\s*)- \[.\].* \[`(.*)`\]\((.*)\).*',
+                'human': r'^(\s*)- \[.\].* `(.*)`: <strong>(.*)</strong>$',
+                'single': r'^(\s*)- \[.\].* `(.*)`: `(.*)`$',
+                'multi': r'^(\s*)- \[.\].* `(.*)`:$',
+                'exitcode': r'^(\s*)- \[.\].* `(.*)`$',
+        }
+
+        command_type = None
+        command_match = None
+
+        for type_c, re_c in command_re.items():
+            match = re.match(re_c, line)
+            if match:
+                command_type = type_c
+                command_match = match
+                break
+
+        if command_type is None:
+            logger.error(f'Failed to match run command line: {line}')
+            return
+
+        command = match[2]
+        output = []
+        insert_offset = self.cursor_offset + 1
+        section_index = -1
+
+        if command_type == 'multi':
+            section_index = self.clean_multi_lines_code()
+
+        try:
+            import subprocess
+
+            proc = subprocess.Popen(command, shell=True, universal_newlines=True,
+                                    bufsize=0,
+                                    stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+            while proc.poll() is None:
+                if command_type == 'multi':
+                    while line := proc.stdout.readline():
+                        section_index = self._insert(section_index, line)
+                        output.append(line)
+                        self.writeout()
+                        self._redraw()
+                        self._render()
+                else:
+                    output.extend(proc.stdout.readlines())
```

Comment:
if a `single` command prints multiple lines (over some period), your `output` list ends up with every line, appended multiple times. Shouldn't matter since we only use the first line but this is still not ideal. I recon this needs to be something like:
```
while line := ...
     output.append(line)
     if multi:
        other stuff
```

also - in the unlikely case that the command finishes before `proc.poll()`, we never gather the output

Reply:
I think I fixed the "in the unlikely case that the command finishes before proc.poll(), we never gather the output".

I really don't understand the first part of the comment.

Basically what the code is supposed to do is that if the output has more than one line, it converts it into a `multi` line, dropping the single line output in the command prompt.

Reply:
> Basically what the code is supposed to do is that if the output has more than one line, it converts it into a multi line, dropping the single line output in the command prompt.

actually, I think was wrong here anyway and this is accommodated for by `proc.stdout` not `lseek`ing between `readlines()` calls. For some reason I thought `readlines()` would always give you *all* lines from the beginning of the file. But that's wrong since nothing `seek`s back to the start - so - nevermind, sorry about that.

Having said that, to test the above I wrote a minimal test script and it didn't behave at all what I expected.  First, `proc.stdout.readlines()` blocks, so this loop can't be entered more than once anyway. Second, the [docs](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.stderr) have a warning that we can deadlock if the pipe is full. Unlikely for our use-case, but it suggests to use `popen.communicate` instead.
And `Popen` is a context manager, so I think something like this should be possible:

```python
stdout, stderr = None, None
proc = subprocess.Popen(...)
with proc:
    try:
        stdout, stderr = popen.communicate(timeout=30)
    except (subprocess.TimeoutExpired, KeyboardInterrupt):
        # Note: communicate() timeout expiry does not terminate the process
        proc.terminate()
        time.sleep(0.5)
        if proc.poll() is None:
            proc.kill()

        stdout, stderr = popen.communicate()

# now handle stdout, stderr and proc.result
```

I think this may make the code look a bit nicer too.